{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alvin/anaconda3/lib/python3.8/site-packages/ete3-3.1.2-py3.7.egg/ete3/evol/parser/codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "/home/alvin/anaconda3/lib/python3.8/site-packages/ete3-3.1.2-py3.7.egg/ete3/evol/parser/codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n"
     ]
    }
   ],
   "source": [
    "from os.path import expanduser\n",
    "from importlib.machinery import SourceFileLoader\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "import itertools \n",
    "\n",
    "import multiprocessing as mp \n",
    "import ete3\n",
    "import subprocess\n",
    "\n",
    "from matplotlib import cm\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "\n",
    "# load custom flu and ngs libraries \n",
    "laeb_lib = expanduser(\"../PYTHONLIB\") # folder where custom libraries are saved \n",
    "fc = SourceFileLoader('fc', \"%s/flu_common.py\"%(laeb_lib)).load_module()\n",
    "ngs = SourceFileLoader('ngs', \"%s/laeb_ngs_pipeline.py\"%(laeb_lib)).load_module()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputs \n",
    "# file path to data folder - fastq files to be analysed must be in {data folder}/raw\n",
    "data_folder = './data' \n",
    "# reference fasta file name (should be placed in input_folder)\n",
    "ref_fasta_fname = './input/H3N2_Bris07.fasta' \n",
    "# CSV file containing the CDR regions of each gene segment (numbering should be based on that of the given reference sequence)\n",
    "cds_coords = \"./input/CDS_H3N2_Bris07.csv\"\n",
    "# file path to metadata file. \n",
    "meta_fname = './results/metadata_w_covbreadth.csv' \n",
    "\n",
    "threadnum = 4 # number of CPU threads for parallelization \n",
    "\n",
    "# variant calling options\n",
    "min_cov = 100 # minimum coverage \n",
    "min_var_freq = 0 # minimum number of variants to be called \n",
    "min_var_prop = 0.02 # minimum variant proportion \n",
    "err_tol = 0.01 # threshold to which variant called could result from base calling error \n",
    "min_breadth = 0.7 # min breadth of gene segment to be mapped for further analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Initialising CDS coordinates...\n",
      "\n",
      "Check translated protein sequences...\n",
      "PB2 MERIKELRNLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPSLRMKWMMAMKYPITADKRITEMVPERNEQGQTLWSKMSDAGSDRVMVSPLAVTWWNRNGPVTSTVHYPKVYKTYFDKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITKEKKEELRDCKISPLMVAYMLERELVRKTRFLPVAGGTSSIYIEVLHLTQGTCWEQMYTPGGGVRNDDVDQSLIIAARNIVRRAAVSADPLASLLEMCHSTQIGGTRMVDILRQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKKEEEVLTGNLQTLKIRVHEGYEEFTMVGKRATAILRKATRRLVQLIVSGRDEQSIAEAIIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGVEHIDSVMGMIGVLPDMTPSTEMSMRGIRVSKMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTERLTITYSSSMMWEINGPESVLVNTYQWIIRNWEAVKIQWSQNPAMLYNKMEFEPFQSLVPKAIRSQYSGFVRTLFQQMRDVLGTFDTTQIIKLLPFAAAPPKQSRMQFSSLTVNVRGSGMRILVRGNSPVFNYNKTTKRLTILGKDAGTLIEDPDESTSGVESAVLRGFLIIGKEDRRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*\n",
      "PB1 MDVNPTLLFIKVPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLETMEAVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTIEVFRSNGLTANESGRLIDFLKDVMESMDKEEMEITTHFQRKRRVRDNMTKKMVTQRTIGKKKQRVNKRGYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITKNQPEWFRNILSIAPIMFSNKMARLGKGYMFESKRMKLRTQIPAEMLASVDLKYFNESTRKKIEKIRPLLIDGTASLSPGMMMGMFNMLSTVLGVSILNLGQKKYTKTTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINKTGTFEFTSFFYRYGFVANFSMELPSFGVSGINESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWDQTQSRAGLLVSDGGPNLYNIRNLHIPEVCLKWELMDENYRGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCNLFEKFFPSSSYRRPIGISSMVEAMVSRARIDARIDFESGRIKKEEFSEIMKICSTIEELRRQK*\n",
      "PB1-F2 MEQEQGTPWTQSTEHTNIQRKGSGRQIQKLGHPNSTQLMDHYLRIMNQVDMHKQTVSWRLWPSLKNPTQVSLRTHALKQWKPFNRQGWTN*\n",
      "PA MEDFVRQCFNPMIVELAEKAMKEYGEDLKIETNKFAAICTHLEVCFMYSDFHFINEQGESIVVELDDPNALLKHRFEIIEGRDRTMAWTVVNSICNTTGAGKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSENTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMANRGLWDSFRQSERGEETIEEKFEITGTMRRLADQSLPPNFSCLENFRAYVDGFEPNGCIEGKLSQMSKEVNAQIEPFLKTTPRPIKLPNGPPCYQRSKFLLMDALKLSIEDPSHEGEGIPLYDAIKCIKTFFGWKEPYIVKPHEKGINSNYLLSWKQVLSELQDIENEEKIPRTKNMKKTSQLKWALGENMAPEKVDFENCRDISDLKQYDSDEPELRSLSSWIQNEFNKACELTDSVWIELDEIGEDVAPIEHIASMRRNYFTAEVSHCRATEYIMKGVYINTALLNASCAAMDDFQLIPMISKCRTKEGRRKTNLYGFIIKGRSHLRNDTDVVNFVSMEFSLTDPRLEPHKWEKYCVLEIGDMLLRNAIGQISRPMFLYVRTNGTSKVKMKWGMEMRRCLLQSLQQIESMIEAESSVKEKDMTKEFFENKSEAWPIGESPKGVEEGSIGKVCRTLLAKSVFNSLYASPQLEGFSAESRKLLLVVQALRDNLEPGTFDLGGLYEAIEECLINDPWVLLNASWFNSFLTHALK*\n",
      "PA-X MEDFVRQCFNPMIVELAEKAMKEYGEDLKIETNKFAAICTHLEVCFMYSDFHFINEQGESIVVELDDPNALLKHRFEIIEGRDRTMAWTVVNSICNTTGAGKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSENTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMANRGLWDSFVSPKEAKKQLKKNLKSQELCVGLPTKVSHRTSPALRILEPMWMDSNRTAALRASFLKCPKK*\n",
      "HA MKTIIALSYILCLVFTQKLPGNDNSTATLCLGHHAVPNGTIVKTITNDQIEVTNATELVQSSSTGEICDSPHQILDGENCTLIDALLGDPQCDGFQNKKWDLFVERSKAYSNCYPYDVPDYASLRSLVASSGTLEFNNESFNWTGVTQNGTSSACIRRSNNSFFSRLNWLTHLKFKYPALNVTMPNNEKFDKLYIWGVHHPGTDNDQIFLYAQASGRITVSTKRSQQTVIPNIGSRPRVRNIPSRISIYWTIVKPGDILLINSTGNLIAPRGYFKIRSGKSSIMRSDAPIGKCNSECITPNGSIPNDKPFQNVNRITYGACPRYVKQNTLKLATGMRNVPEKQTRGIFGAIAGFIENGWEGMVDGWYGFRHQNSEGIGQAADLKSTQAAIDQINGKLNRLIGKTNEKFHQIEKEFSEVEGRIQDLEKYVEDTKIDLWSYNAELLVALENQHTIDLTDSEMNKLFEKTKKQLRENAEDMGNGCFKIYHKCDNACIGSIRNGTYDHDVYRDEALNNRFQIKGVELKSGYKDWILWISFAISCFLLCVALLGFIMWACQKGNIRCNICI*\n",
      "NP MASQGTKRSYEQMETDGDRQNATEIRASVGKMIDGIGRFYIQMCTELKLSDHEGRLIQNSLTIEKMVLSAFDERRNKYLEEHPSAGKDPKKTGGPIYRRVDGKWMRELVLYDKEEIRRIWRQANNGEDATSGLTHIMIWHSNLNDATYQRTRALVRTGMDPRMCSLMQGSTLPRRSGAAGAAVKGIGTMVMELIRMVKRGINDRNFWRGENGRKTRSAYERMCNILKGKFQTAAQRAMVDQVRESRNPGNAEIEDLIFLARSALILRGSVAHKSCLPACAYGPAVSSGYDFEKEGYSLVGIDPFKLLQNSQIYSLIRPNENPAHKSQLVWMACHSAAFEDLRLLSFIRGTKVSPRGKLSTRGVQIASNENMDNMGSSTLELRSGYWAIRTRSGGNTNQQRASAGQTSVQPTFSVQRNLPFEKSTIMAAFTGNTEGRTSDMRAEIIRMMEGAKPEEVSFRGRGVFELSDEKATNPIVPSFDMSNEGSYFFGDNAEEYDN*\n",
      "NA MNPNQKIITIGSVSLTISTICFFMQIAILITTVTLHFKQYEFNSPPNNQVMLCEPTIIERNITEIVYLTNTTIEKEICPKLAEYRNWSKPQCDITGFAPFSKDNSIRLSAGGDIWVTREPYVSCDPDKCYQFALGQGTTLNNVHSNDTVRDRTPYRTLLMNELGVPFHLGTKQVCIAWSSSSCHDGKAWLHVCITGDDKNATASFIYNGRLVDSIVSWSKEILRTQESECVCINGTCTVVMTDGSASGKADTKILFIEEGKIVHTSTLSGSAQHVEECSCYPRYPGVRCVCRDNWKGSNRPIVDINIKDHSTVSSYVCSGLVGDTPRKNDSSSSSHCLDPNNEEGGHGVKGWAFDDGNDVWMGRTISEKSRLGYETFKVIEGWSNPKSKLQINRQVIVDRGNRSGYSGIFSVEGKSCINRCFYVELIRGRKEETEVLWTSNSIVVFCGTSGTYGTGSWPDGADINLMPI*\n",
      "M1 MSLLTEVETYVLSIVPSGPLKAEIAQRLEDVFAGKNTDLEALMEWLKTRPILSPLTKGILGFVFTLTVPSERGLQRRRFVQNALNGNGDPNNMDKAVKLYRKLKREITFHGAKEIALSYSAGALASCMGLIYNRMGAVTTEVAFGLVCATCEQIADSQHRSHRQMVATTNPLIRHENRMVLASTTAKAMEQMAGSSEQAAEAMEIASQARQMVQAMRAIGTHPSSSTGLRDDLLENLQTYQKRMGVQMQRFK*\n",
      "M2 MSLLTEVETPIRNEWGCRCNDSSDPLVVAANIIGILHLILWILDRLFFKCVYRLFKHGLKRGPSTEGVPESMREEYRKEQQNAVDADDSHFVSIELE*\n",
      "NEP MDSNTVSSFQCLLFQDILLRMSKMQLGSSSEDLNGMITQFESLKIYRDSLGEAVMRMGDLHLLQNRNGKWREQLGQKFEEIRWLIEEVRHRLKTTENSFEQITFMQALQLLFEVEQEIRTFSFQLI*\n",
      "NS1 MDSNTVSSFQVDCFLWHIRKQVVDQELSDAPFLDRLRRDQRSLRGRGNTLGLDIKAATHVGKQIVEKILKEESDEALKMTMVSTPASRYITDMTIEELSRNWFMLMPKQKVEGPLCIRMDQAIMEKNIMLKANFSVIFDRLETIVLLRAFTEEGAIVGEISPLPSFPGHTIEDVKNAIGVLIGGLEWNDNTVRVSKNLQRFAWRSSNENGGPPLTPKQKREMARTARSKV*\n"
     ]
    }
   ],
   "source": [
    "# presets \n",
    "reffasta = ref_fasta_fname\n",
    "\n",
    "# initialise\n",
    "gene_to_proteinorf, influenza_gene_len, sorted_refnames, nucpos_shift = ngs.initialisation(cds_coords, reffasta, laeb_lib)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata and variant calling files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>enrolD</th>\n",
       "      <th>idx</th>\n",
       "      <th>ct</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>aliquot</th>\n",
       "      <th>run</th>\n",
       "      <th>primer_i5</th>\n",
       "      <th>primer_i7</th>\n",
       "      <th>enrol-onset</th>\n",
       "      <th>...</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>patch</th>\n",
       "      <th>1-PB2</th>\n",
       "      <th>2-PB1</th>\n",
       "      <th>3-PA</th>\n",
       "      <th>4-HA</th>\n",
       "      <th>5-NP</th>\n",
       "      <th>6-NA</th>\n",
       "      <th>7-M</th>\n",
       "      <th>8-NS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3N2_1205_T0_T_S517_N701_R</th>\n",
       "      <td>1205</td>\n",
       "      <td>0</td>\n",
       "      <td>R1</td>\n",
       "      <td>26</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S517</td>\n",
       "      <td>N701</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.945946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1205_T1_N_S517_N702_R</th>\n",
       "      <td>1205</td>\n",
       "      <td>1</td>\n",
       "      <td>R2</td>\n",
       "      <td>25</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S517</td>\n",
       "      <td>N702</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931034</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1205_T2_T_S517_N703_R</th>\n",
       "      <td>1205</td>\n",
       "      <td>2</td>\n",
       "      <td>R3</td>\n",
       "      <td>30</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S517</td>\n",
       "      <td>N703</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.907407</td>\n",
       "      <td>0.985294</td>\n",
       "      <td>0.866667</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1205_T3_T_S517_N704_R</th>\n",
       "      <td>1205</td>\n",
       "      <td>3</td>\n",
       "      <td>R4</td>\n",
       "      <td>29</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S517</td>\n",
       "      <td>N704</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1205_T4_T_S517_N705_R</th>\n",
       "      <td>1205</td>\n",
       "      <td>4</td>\n",
       "      <td>R5</td>\n",
       "      <td>28</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S517</td>\n",
       "      <td>N705</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            subject_id  enrolD idx  ct SampleType  aliquot  \\\n",
       "sampid                                                                       \n",
       "H3N2_1205_T0_T_S517_N701_R        1205       0  R1  26          T        1   \n",
       "H3N2_1205_T1_N_S517_N702_R        1205       1  R2  25          N        1   \n",
       "H3N2_1205_T2_T_S517_N703_R        1205       2  R3  30          T        1   \n",
       "H3N2_1205_T3_T_S517_N704_R        1205       3  R4  29          T        1   \n",
       "H3N2_1205_T4_T_S517_N705_R        1205       4  R5  28          T        1   \n",
       "\n",
       "                            run primer_i5 primer_i7  enrol-onset  ...  \\\n",
       "sampid                                                            ...   \n",
       "H3N2_1205_T0_T_S517_N701_R    1      S517      N701          6.0  ...   \n",
       "H3N2_1205_T1_N_S517_N702_R    1      S517      N702          6.0  ...   \n",
       "H3N2_1205_T2_T_S517_N703_R    1      S517      N703          6.0  ...   \n",
       "H3N2_1205_T3_T_S517_N704_R    1      S517      N704          6.0  ...   \n",
       "H3N2_1205_T4_T_S517_N705_R    1      S517      N705          6.0  ...   \n",
       "\n",
       "                           timepoint  patch  1-PB2  2-PB1      3-PA      4-HA  \\\n",
       "sampid                                                                          \n",
       "H3N2_1205_T0_T_S517_N701_R       6.0      0    1.0    1.0  0.962963  1.000000   \n",
       "H3N2_1205_T1_N_S517_N702_R       7.0      0    1.0    1.0  1.000000  1.000000   \n",
       "H3N2_1205_T2_T_S517_N703_R       8.0      0    1.0    1.0  0.907407  0.985294   \n",
       "H3N2_1205_T3_T_S517_N704_R       9.0      0    1.0    1.0  0.962963  1.000000   \n",
       "H3N2_1205_T4_T_S517_N705_R      10.0      0    1.0    1.0  0.962963  1.000000   \n",
       "\n",
       "                                5-NP      6-NA  7-M      8-NS  \n",
       "sampid                                                         \n",
       "H3N2_1205_T0_T_S517_N701_R  1.000000  0.844828  1.0  0.945946  \n",
       "H3N2_1205_T1_N_S517_N702_R  1.000000  0.931034  1.0  1.000000  \n",
       "H3N2_1205_T2_T_S517_N703_R  0.866667  0.844828  1.0  1.000000  \n",
       "H3N2_1205_T3_T_S517_N704_R  0.966667  0.844828  1.0  1.000000  \n",
       "H3N2_1205_T4_T_S517_N705_R  1.000000  0.844828  1.0  1.000000  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>nuc_var</th>\n",
       "      <th>nuc_prop</th>\n",
       "      <th>nuc_freq</th>\n",
       "      <th>nuc_consensus</th>\n",
       "      <th>nuc_consensus_freq</th>\n",
       "      <th>nuc_coverage</th>\n",
       "      <th>protein</th>\n",
       "      <th>aapos</th>\n",
       "      <th>HA_num_type</th>\n",
       "      <th>aa_var</th>\n",
       "      <th>aa_prop</th>\n",
       "      <th>aa_freq</th>\n",
       "      <th>expected_aa_consensus</th>\n",
       "      <th>aa_consensus</th>\n",
       "      <th>aa_consensus_freq</th>\n",
       "      <th>aa_coverage</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>codon</th>\n",
       "      <th>nonsyn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th>gene</th>\n",
       "      <th>nucpos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">H3N2_1205_T0_T_S517_N701_R</th>\n",
       "      <th>1-PB2</th>\n",
       "      <th>1221</th>\n",
       "      <td>A</td>\n",
       "      <td>0.080098</td>\n",
       "      <td>293</td>\n",
       "      <td>G</td>\n",
       "      <td>3358</td>\n",
       "      <td>3658</td>\n",
       "      <td>PB2</td>\n",
       "      <td>407</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>0.082117</td>\n",
       "      <td>287.0</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>3196.0</td>\n",
       "      <td>3495.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GAA</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">4-HA</th>\n",
       "      <th>939</th>\n",
       "      <td>G</td>\n",
       "      <td>0.073570</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>3336</td>\n",
       "      <td>3602</td>\n",
       "      <td>HA</td>\n",
       "      <td>297</td>\n",
       "      <td>H3</td>\n",
       "      <td>V</td>\n",
       "      <td>0.074179</td>\n",
       "      <td>262.0</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>3267.0</td>\n",
       "      <td>3532.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GTG</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>C</td>\n",
       "      <td>0.044221</td>\n",
       "      <td>88</td>\n",
       "      <td>T</td>\n",
       "      <td>1902</td>\n",
       "      <td>1990</td>\n",
       "      <td>HA</td>\n",
       "      <td>409</td>\n",
       "      <td>H3</td>\n",
       "      <td>L</td>\n",
       "      <td>0.042916</td>\n",
       "      <td>83.0</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>1839.0</td>\n",
       "      <td>1934.0</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">5-NP</th>\n",
       "      <th>1058</th>\n",
       "      <td>T</td>\n",
       "      <td>0.032740</td>\n",
       "      <td>125</td>\n",
       "      <td>C</td>\n",
       "      <td>3689</td>\n",
       "      <td>3818</td>\n",
       "      <td>NP</td>\n",
       "      <td>353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>0.033270</td>\n",
       "      <td>122.0</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>3531.0</td>\n",
       "      <td>3667.0</td>\n",
       "      <td>2</td>\n",
       "      <td>TTT</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1150</th>\n",
       "      <td>A</td>\n",
       "      <td>0.123601</td>\n",
       "      <td>276</td>\n",
       "      <td>G</td>\n",
       "      <td>1955</td>\n",
       "      <td>2233</td>\n",
       "      <td>NP</td>\n",
       "      <td>384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>0.124709</td>\n",
       "      <td>268.0</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>1871.0</td>\n",
       "      <td>2149.0</td>\n",
       "      <td>1</td>\n",
       "      <td>AGG</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        nuc_var  nuc_prop  nuc_freq  \\\n",
       "sampid                     gene  nucpos                               \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221         A  0.080098       293   \n",
       "                           4-HA  939          G  0.073570       265   \n",
       "                                 1275         C  0.044221        88   \n",
       "                           5-NP  1058         T  0.032740       125   \n",
       "                                 1150         A  0.123601       276   \n",
       "\n",
       "                                        nuc_consensus  nuc_consensus_freq  \\\n",
       "sampid                     gene  nucpos                                     \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221               G                3358   \n",
       "                           4-HA  939                A                3336   \n",
       "                                 1275               T                1902   \n",
       "                           5-NP  1058               C                3689   \n",
       "                                 1150               G                1955   \n",
       "\n",
       "                                         nuc_coverage protein  aapos  \\\n",
       "sampid                     gene  nucpos                                \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221            3658     PB2    407   \n",
       "                           4-HA  939             3602      HA    297   \n",
       "                                 1275            1990      HA    409   \n",
       "                           5-NP  1058            3818      NP    353   \n",
       "                                 1150            2233      NP    384   \n",
       "\n",
       "                                        HA_num_type aa_var   aa_prop  aa_freq  \\\n",
       "sampid                     gene  nucpos                                         \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221           NaN      E  0.082117    287.0   \n",
       "                           4-HA  939             H3      V  0.074179    262.0   \n",
       "                                 1275            H3      L  0.042916     83.0   \n",
       "                           5-NP  1058           NaN      F  0.033270    122.0   \n",
       "                                 1150           NaN      R  0.124709    268.0   \n",
       "\n",
       "                                        expected_aa_consensus aa_consensus  \\\n",
       "sampid                     gene  nucpos                                      \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221                       E            E   \n",
       "                           4-HA  939                        V            V   \n",
       "                                 1275                       L            L   \n",
       "                           5-NP  1058                       S            S   \n",
       "                                 1150                       G            G   \n",
       "\n",
       "                                         aa_consensus_freq  aa_coverage  \\\n",
       "sampid                     gene  nucpos                                   \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221               3196.0       3495.0   \n",
       "                           4-HA  939                3267.0       3532.0   \n",
       "                                 1275               1839.0       1934.0   \n",
       "                           5-NP  1058               3531.0       3667.0   \n",
       "                                 1150               1871.0       2149.0   \n",
       "\n",
       "                                         codon_pos codon  nonsyn  \n",
       "sampid                     gene  nucpos                           \n",
       "H3N2_1205_T0_T_S517_N701_R 1-PB2 1221            3   GAA     1.0  \n",
       "                           4-HA  939             3   GTG     1.0  \n",
       "                                 1275            3   CTC     1.0  \n",
       "                           5-NP  1058            2   TTT     0.0  \n",
       "                                 1150            1   AGG     0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# metadata \n",
    "## metadata must have 'sampid' header which is used as sample identifier \n",
    "meta_df = pd.read_csv('%s'%meta_fname)\n",
    "sorted_sampid = sorted(set(meta_df['sampid']))\n",
    "\n",
    "meta_df = meta_df.set_index(\"sampid\")\n",
    "display (meta_df.head())\n",
    "\n",
    "# vcf \n",
    "variant_call_df = pd.read_csv(\"./results/variant_call_MinCoV%i_MinProp%.2f_MinFreq%i_ErrTol%.2f.csv\"%(min_cov, min_var_prop, min_var_freq, err_tol))\n",
    "variant_call_df = variant_call_df.set_index([\"sampid\", 'gene', \"nucpos\"])\n",
    "display (variant_call_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get consensus and putative minority haplotype sequences for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th>gene</th>\n",
       "      <th>seqtype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">H3N2_1205_T0_T_S517_N701_R</th>\n",
       "      <th>3-PA</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGGAAGATTTTGTGCGACAATGCTTCAACCCGATGATTGTCGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">4-HA</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>ATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">5-NP</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGGCGTCCCAAGGCACCAAACGGTCTTATGAACAGATGGAAACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>ATGGCGTCCCAAGGCACCAAACGGTCTTATGAACAGATGGAAACTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                       seq\n",
       "sampid                     gene seqtype                                                   \n",
       "H3N2_1205_T0_T_S517_N701_R 3-PA C        ATGGAAGATTTTGTGCGACAATGCTTCAACCCGATGATTGTCGAAC...\n",
       "                           4-HA C        ATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCG...\n",
       "                                M1       ATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCG...\n",
       "                           5-NP C        ATGGCGTCCCAAGGCACCAAACGGTCTTATGAACAGATGGAAACTG...\n",
       "                                M1       ATGGCGTCCCAAGGCACCAAACGGTCTTATGAACAGATGGAAACTG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if not os.path.isdir(\"./reference\"): \n",
    "    os.mkdir(\"./reference\")\n",
    "\n",
    "def get_consensus_minority_seq(sampid):\n",
    "    gene_to_consensus_seq_row = []\n",
    "    \n",
    "    # check that all gene segment satisfy min_breadth \n",
    "    if (meta_df.loc[sampid, sorted_refnames]>min_breadth).all() == False: \n",
    "        return gene_to_consensus_seq_row\n",
    "\n",
    "    # read map_nuc_results \n",
    "    map_nuc_results = pd.read_csv('./results/map_nuc_results_%s.csv'%(sampid))\n",
    "    map_nuc_results = map_nuc_results.set_index([\"Gene\", \"Position\"])\n",
    "\n",
    "    # sample_vcf \n",
    "    try:\n",
    "        sample_vcf = variant_call_df.loc[sampid]\n",
    "    except: \n",
    "        sample_vcf = None\n",
    "\n",
    "    for gene in set(map_nuc_results.index.get_level_values(0)): \n",
    "\n",
    "        # add consensus sequence \n",
    "        sequence = {position:map_nuc_results.loc[(gene, position), \"Consensus\"] if (gene, position) in list(map_nuc_results.index) else \"-\" for position in range(1, influenza_gene_len[gene]+1)}\n",
    "        conseq = \"\".join([sequence[pos] for pos in sorted(sequence.keys())])\n",
    "        gene_to_consensus_seq_row.append({\"sampid\":sampid, \"gene\":gene, \"seq\":conseq, \"seqtype\":\"C\"})\n",
    "\n",
    "        # add putative minority haplotype sequence \n",
    "        try: \n",
    "            gene_sample_vcf = sample_vcf.loc[gene].copy()\n",
    "        except: \n",
    "            continue \n",
    "\n",
    "        if isinstance(gene_sample_vcf, pd.Series):\n",
    "            gene_sample_vcf = gene_sample_vcf.to_frame().T\n",
    "\n",
    "        gene_sample_vcf = gene_sample_vcf.drop_duplicates([\"nucpos\", \"nuc_var\"])\n",
    "        gene_sample_vcf = gene_sample_vcf.set_index(['nucpos'])\n",
    "        \n",
    "        pos_to_var = {}\n",
    "        for pos in list(gene_sample_vcf.index):\n",
    "            nuc_var = gene_sample_vcf.loc[pos, \"nuc_var\"]\n",
    "            if isinstance(nuc_var, str): # single variant \n",
    "                pos_to_var[pos] = [gene_sample_vcf.loc[pos, \"nuc_var\"]]\n",
    "            else: # multi variant \n",
    "                pos_to_var[pos] = list(gene_sample_vcf.loc[pos, 'nuc_var'])\n",
    "        \n",
    "        minseq = [pos_to_var[pos] if pos in pos_to_var else [sequence[pos]] for pos in range(1, influenza_gene_len[gene]+1)]\n",
    "        for midx, mseq in enumerate(list(itertools.product(*minseq))): \n",
    "            gene_to_consensus_seq_row.append({\"sampid\":sampid, \"gene\":gene, \"seq\":\"\".join(mseq), \"seqtype\":\"M%i\"%(midx+1)})\n",
    "        \n",
    "    return gene_to_consensus_seq_row\n",
    "\n",
    "if os.path.isfile(\"./reference/consensus_sequences.csv\"):\n",
    "    gene_to_consensus_seq = pd.read_csv(\"./reference/consensus_sequences.csv\")\n",
    "else:\n",
    "    pool = mp.Pool(processes=threadnum)\n",
    "    results = [pool.apply_async(get_consensus_minority_seq, args=(sampid,)) for sampid in sorted_sampid]\n",
    "    output = [p.get() for p in results]\n",
    "\n",
    "    gene_to_consensus_seq = [k for v in list(filter(None, output)) for k in v]\n",
    "    gene_to_consensus_seq = pd.DataFrame.from_dict(gene_to_consensus_seq)\n",
    "\n",
    "    # save to file \n",
    "    gene_to_consensus_seq.to_csv(\"./reference/consensus_sequences.csv\", index=False)\n",
    "\n",
    "gene_to_consensus_seq = gene_to_consensus_seq.set_index([\"sampid\", \"gene\", \"seqtype\"])\n",
    "display (gene_to_consensus_seq.head()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenate consensus sequences of all gene segments "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome = fc.parsefasta(reffasta)\n",
    "\n",
    "# write fasta \n",
    "if os.path.isfile(\"./reference/concatenated_wgs.fasta\") == False:\n",
    "    with open(\"./reference/concatenated_wgs.fasta\", \"w\") as output: \n",
    "        output.write(\">%s\\n%s\\n\"%(re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1]), \n",
    "                                  re.sub(\"TGATAAAAAACACCCTTGTTTCTAC$\", \"\", \"\".join([reference_genome[gene] for gene in sorted_refnames]).upper())))\n",
    "\n",
    "        for sampid in set(gene_to_consensus_seq.index.get_level_values(0)): \n",
    "            #  write concatenated consensus sequence \n",
    "            sample_gene_to_consensus_seq = gene_to_consensus_seq.loc[sampid]\n",
    "\n",
    "            header = \"%s|%s\"%(sampid, \"C\")\n",
    "            sequence =  \"\".join([sample_gene_to_consensus_seq.loc[(gene, \"C\"), \"seq\"] for gene in sorted_refnames])\n",
    "            output.write(\">%s\\n%s\\n\"%(header, sequence))\n",
    "\n",
    "            # write putative minority haplotype sequence \n",
    "            for seqtype in set(sample_gene_to_consensus_seq.index.get_level_values(1)): \n",
    "                if seqtype == \"C\": \n",
    "                    continue \n",
    "\n",
    "                header = \"%s|%s\"%(sampid, seqtype)\n",
    "                sequence =  []\n",
    "                for gene in sorted_refnames: \n",
    "                    if (gene, seqtype) in sample_gene_to_consensus_seq.index: \n",
    "                        sequence.append(sample_gene_to_consensus_seq.loc[(gene, seqtype), \"seq\"])\n",
    "                    else: \n",
    "                        sequence.append(sample_gene_to_consensus_seq.loc[(gene, \"C\"), \"seq\"])\n",
    "                output.write(\">%s\\n%s\\n\"%(header, \"\".join(sequence))) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruct phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build tree \n",
    "if os.path.isfile(\"./reference/concatenated_wgs.fasta.treefile\") == False:\n",
    "    cmd = ['iqtree', '-nt', 'AUTO', '-ntmax', str(threadnum), \n",
    "           '-m', 'GTR+I+G4', \n",
    "           '-s', './reference/concatenated_wgs.fasta']\n",
    "    subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate newick tree for printing and annotation dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# generate ggtree dataframe and newick tree file\n",
    "# get annotations \n",
    "tree_annotations = []\n",
    "\n",
    "ete3_tree = ete3.Tree(\"./reference/concatenated_wgs.fasta.treefile\")\n",
    "ete3_tree.set_outgroup(re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1]))\n",
    "ete3_tree.ladderize()\n",
    "\n",
    "for leaf in ete3_tree.get_leaves():\n",
    "    # skip vaccine root \n",
    "    if (leaf.name == \"H3N2_Bris07\"): \n",
    "        print_name = leaf.name\n",
    "        subject_id = \"\"\n",
    "        seq_type = \"Consensus\" \n",
    "    else: \n",
    "        subject_id = re.search(\"H3N2_(\\d+)\", leaf.name).group(1)\n",
    "        sampid = re.sub(\"\\|.+$\", \"\", leaf.name)\n",
    "\n",
    "        timepoint = meta_df.loc[sampid, \"timepoint\"]\n",
    "        if pd.isna(timepoint): \n",
    "            timepoint = \"D-NaN\"\n",
    "        else: \n",
    "            timepoint = \"D%i\"%(int(timepoint))\n",
    "\n",
    "        if re.search(\"C$\", leaf.name): \n",
    "            seq_type = \"Consensus\"\n",
    "        else: \n",
    "            if subject_id == \"1707\":\n",
    "                minority_idx = int(re.search(\"M(\\d+)$\", leaf.name).group(1))\n",
    "                seq_type = \"Minor-%i\"%(minority_idx)\n",
    "            else:\n",
    "                seq_type = \"Minor\"\n",
    "\n",
    "        print_name = \"_\".join([subject_id, timepoint, seq_type])\n",
    "        \n",
    "    tree_annotations.append({\"index\":leaf.name, \"print_name\":print_name, \n",
    "                                 \"show_name\":1, \n",
    "                                 \"show_tip\":1 if subject_id in [\"1673\", \"1878\"] else 0, \n",
    "                                 \"subject_id\":subject_id})\n",
    "\n",
    "# save for ggtree use \n",
    "tree_annotations = pd.DataFrame.from_dict(tree_annotations).set_index(\"index\")\n",
    "tree_annotations.to_csv(\"./reference/ggtree_concatenated_wgs.meta.csv\")\n",
    "ete3_tree.write(outfile=\"./reference/ggtree_concatenated_wgs.nwk\", format=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse concatenated genome tree and identify samples that may have derived from mixed infections/contaminated sammples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1418 0\n",
      "1673 0\n",
      "1682 1\n",
      "1878 1\n"
     ]
    }
   ],
   "source": [
    "tree = ete3.Tree('./reference/concatenated_wgs.fasta.treefile', format=1)\n",
    "tree.set_outgroup(re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1]))\n",
    "tree.ladderize()\n",
    "\n",
    "subject_to_ca = {}\n",
    "subject_to_headers = []\n",
    "sampid_mixed_infections = {}\n",
    "\n",
    "for subject_id in set(meta_df['subject_id']): \n",
    "\n",
    "    subject_meta_df = meta_df[meta_df['subject_id']==subject_id]\n",
    "    subject_sampid_analysed = list(set(subject_meta_df.index)&set(gene_to_consensus_seq.index.get_level_values(0)))\n",
    "    if (len(subject_sampid_analysed)) == 0:\n",
    "        continue \n",
    "    \n",
    "    all_subject_headers = []\n",
    "    for sampid in subject_sampid_analysed:\n",
    "        sample_gene_to_consensus_seq = gene_to_consensus_seq.loc[sampid]\n",
    "        all_subject_headers += [\"%s|%s\"%(sampid, seqtype) for seqtype in set(sample_gene_to_consensus_seq.index.get_level_values(1))]\n",
    "    \n",
    "    for header in all_subject_headers: \n",
    "        subject_to_headers.append({'subject_id':subject_id, 'header':header})\n",
    "    \n",
    "    ca = tree.get_common_ancestor(all_subject_headers)\n",
    "    subject_to_ca[subject_id] = ca\n",
    "\n",
    "    \n",
    "    # all samples from subject are completely monophyletic \n",
    "    if set(all_subject_headers) == set(ca.get_leaf_names()): \n",
    "        continue \n",
    "    else: \n",
    "        # possibly mixed \n",
    "        sampid_mixed_infections[subject_id] = ca.get_leaf_names()\n",
    "\n",
    "subject_to_headers = pd.DataFrame.from_dict(subject_to_headers)\n",
    "\n",
    "for subject_id, clade_headers in sampid_mixed_infections.items():\n",
    "    all_subject_headers = list(subject_to_headers[subject_to_headers['subject_id']==subject_id]['header'])\n",
    "    other_subject_headers = list(set(clade_headers)-set(all_subject_headers))\n",
    "    subject_ca = subject_to_ca[subject_id]\n",
    "    \n",
    "    mixed_boolean = 0\n",
    "    for other_subject in list(set(subject_to_headers[subject_to_headers['header'].isin(other_subject_headers)]['subject_id'])):\n",
    "        if subject_to_ca[other_subject] not in subject_ca.get_descendants():\n",
    "            mixed_boolean = 1\n",
    "    \n",
    "    print (subject_id, mixed_boolean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>enrolD</th>\n",
       "      <th>idx</th>\n",
       "      <th>ct</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>aliquot</th>\n",
       "      <th>run</th>\n",
       "      <th>primer_i5</th>\n",
       "      <th>primer_i7</th>\n",
       "      <th>enrol-onset</th>\n",
       "      <th>...</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>patch</th>\n",
       "      <th>1-PB2</th>\n",
       "      <th>2-PB1</th>\n",
       "      <th>3-PA</th>\n",
       "      <th>4-HA</th>\n",
       "      <th>5-NP</th>\n",
       "      <th>6-NA</th>\n",
       "      <th>7-M</th>\n",
       "      <th>8-NS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3N2_1673_T0_T_S507_N701_R</th>\n",
       "      <td>1673</td>\n",
       "      <td>0</td>\n",
       "      <td>R73</td>\n",
       "      <td>30</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S507</td>\n",
       "      <td>N701</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1673_T1_N_S507_N702_R</th>\n",
       "      <td>1673</td>\n",
       "      <td>1</td>\n",
       "      <td>R74</td>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S507</td>\n",
       "      <td>N702</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.982759</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1673_T2_T_S507_N703_R</th>\n",
       "      <td>1673</td>\n",
       "      <td>2</td>\n",
       "      <td>R75</td>\n",
       "      <td>34</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S507</td>\n",
       "      <td>N703</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            subject_id  enrolD  idx  ct SampleType  aliquot  \\\n",
       "sampid                                                                        \n",
       "H3N2_1673_T0_T_S507_N701_R        1673       0  R73  30          T        1   \n",
       "H3N2_1673_T1_N_S507_N702_R        1673       1  R74  30          N        1   \n",
       "H3N2_1673_T2_T_S507_N703_R        1673       2  R75  34          T        1   \n",
       "\n",
       "                            run primer_i5 primer_i7  enrol-onset  ...  \\\n",
       "sampid                                                            ...   \n",
       "H3N2_1673_T0_T_S507_N701_R    1      S507      N701          3.0  ...   \n",
       "H3N2_1673_T1_N_S507_N702_R    1      S507      N702          3.0  ...   \n",
       "H3N2_1673_T2_T_S507_N703_R    1      S507      N703          3.0  ...   \n",
       "\n",
       "                           timepoint  patch  1-PB2  2-PB1      3-PA      4-HA  \\\n",
       "sampid                                                                          \n",
       "H3N2_1673_T0_T_S507_N701_R       3.0      0    1.0    1.0  0.944444  1.000000   \n",
       "H3N2_1673_T1_N_S507_N702_R       4.0      0    1.0    1.0  1.000000  1.000000   \n",
       "H3N2_1673_T2_T_S507_N703_R       5.0      0    1.0    1.0  0.870370  0.926471   \n",
       "\n",
       "                            5-NP      6-NA  7-M  8-NS  \n",
       "sampid                                                 \n",
       "H3N2_1673_T0_T_S507_N701_R   1.0  0.844828  1.0   1.0  \n",
       "H3N2_1673_T1_N_S507_N702_R   1.0  0.982759  1.0   1.0  \n",
       "H3N2_1673_T2_T_S507_N703_R   1.0  0.844828  1.0   1.0  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1673 H3N2_1673_T0_T_S507_N701_R 39\n",
      "1673 H3N2_1673_T1_N_S507_N702_R 48\n",
      "1673 H3N2_1673_T2_T_S507_N703_R 94\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>enrolD</th>\n",
       "      <th>idx</th>\n",
       "      <th>ct</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>aliquot</th>\n",
       "      <th>run</th>\n",
       "      <th>primer_i5</th>\n",
       "      <th>primer_i7</th>\n",
       "      <th>enrol-onset</th>\n",
       "      <th>...</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>patch</th>\n",
       "      <th>1-PB2</th>\n",
       "      <th>2-PB1</th>\n",
       "      <th>3-PA</th>\n",
       "      <th>4-HA</th>\n",
       "      <th>5-NP</th>\n",
       "      <th>6-NA</th>\n",
       "      <th>7-M</th>\n",
       "      <th>8-NS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3N2_1878_T0_N_S517_N707_R</th>\n",
       "      <td>1878</td>\n",
       "      <td>0</td>\n",
       "      <td>R199</td>\n",
       "      <td>31</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>S517</td>\n",
       "      <td>N707</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1878_T1_N_S517_N708_R</th>\n",
       "      <td>1878</td>\n",
       "      <td>1</td>\n",
       "      <td>R200</td>\n",
       "      <td>25</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>S517</td>\n",
       "      <td>N708</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3N2_1878_T3_N_S517_N709_R</th>\n",
       "      <td>1878</td>\n",
       "      <td>3</td>\n",
       "      <td>R201</td>\n",
       "      <td>33</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>S517</td>\n",
       "      <td>N709</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>0.965517</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            subject_id  enrolD   idx  ct SampleType  aliquot  \\\n",
       "sampid                                                                         \n",
       "H3N2_1878_T0_N_S517_N707_R        1878       0  R199  31          N        1   \n",
       "H3N2_1878_T1_N_S517_N708_R        1878       1  R200  25          N        1   \n",
       "H3N2_1878_T3_N_S517_N709_R        1878       3  R201  33          N        1   \n",
       "\n",
       "                            run primer_i5 primer_i7  enrol-onset  ...  \\\n",
       "sampid                                                            ...   \n",
       "H3N2_1878_T0_N_S517_N707_R    3      S517      N707          5.0  ...   \n",
       "H3N2_1878_T1_N_S517_N708_R    3      S517      N708          5.0  ...   \n",
       "H3N2_1878_T3_N_S517_N709_R    3      S517      N709          5.0  ...   \n",
       "\n",
       "                           timepoint  patch     1-PB2  2-PB1      3-PA  \\\n",
       "sampid                                                                   \n",
       "H3N2_1878_T0_N_S517_N707_R       5.0      0  0.966667    1.0  0.962963   \n",
       "H3N2_1878_T1_N_S517_N708_R       6.0      0  1.000000    1.0  1.000000   \n",
       "H3N2_1878_T3_N_S517_N709_R       8.0      0  0.966667    1.0  0.962963   \n",
       "\n",
       "                                4-HA      5-NP      6-NA  7-M  8-NS  \n",
       "sampid                                                               \n",
       "H3N2_1878_T0_N_S517_N707_R  0.867647  0.750000  1.000000  1.0   1.0  \n",
       "H3N2_1878_T1_N_S517_N708_R  1.000000  1.000000  1.000000  1.0   1.0  \n",
       "H3N2_1878_T3_N_S517_N709_R  1.000000  0.966667  0.965517  1.0   1.0  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1878 H3N2_1878_T0_N_S517_N707_R 9\n",
      "1878 H3N2_1878_T1_N_S517_N708_R 10\n",
      "1878 H3N2_1878_T3_N_S517_N709_R 75\n"
     ]
    }
   ],
   "source": [
    "outlier_subjects = [1673, 1878]\n",
    "for subject_id in outlier_subjects: \n",
    "    display (meta_df[meta_df['subject_id']==subject_id])\n",
    "    for sampid in (meta_df[meta_df['subject_id']==subject_id].index): \n",
    "        print (subject_id, sampid, len(variant_call_df.loc[sampid].reset_index().drop_duplicates(['gene', 'nucpos', 'nuc_var'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rewrite metadata without likely mixed infection/contaminated samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df = meta_df[~meta_df.index.isin(['H3N2_1673_T0_T_S507_N701_R', 'H3N2_1673_T1_N_S507_N702_R', 'H3N2_1673_T2_T_S507_N703_R', 'H3N2_1878_T3_N_S517_N709_R'])]\n",
    "meta_df.to_csv(\"./results/metadata_wo_mixed_infections.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
