{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import libraries "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os.path import expanduser\n",
    "from importlib.machinery import SourceFileLoader\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "import itertools \n",
    "\n",
    "import multiprocessing as mp \n",
    "import ete3\n",
    "import subprocess\n",
    "\n",
    "from matplotlib import cm\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "\n",
    "# load custom flu and ngs libraries \n",
    "laeb_lib = expanduser(\"./python_lib\") # folder where custom libraries are saved \n",
    "fc = SourceFileLoader('fc', \"%s/flu_common.py\"%(laeb_lib)).load_module()\n",
    "ngs = SourceFileLoader('ngs', \"%s/laeb_ngs_pipeline.py\"%(laeb_lib)).load_module()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputs \n",
    "# file path to data folder - fastq files to be analysed must be in {data folder}/raw\n",
    "data_folder = './data' \n",
    "# reference fasta file name (should be placed in input_folder)\n",
    "ref_fasta_fname = './input/H1N1pdm09_Cali09.fasta' \n",
    "# CSV file containing the CDR regions of each gene segment (numbering should be based on that of the given reference sequence)\n",
    "cds_coords = \"./input/CDS_H1N1pdm09_Cali09.csv\"\n",
    "nucpos_shift = \"./input/CDS_shift_H1N1pdm09_Cali09.csv\"\n",
    "# file path to metadata file. \n",
    "meta_fname = './results/metadata_w_covbreadth.csv' \n",
    "\n",
    "threadnum = 4 # number of CPU threads for parallelization \n",
    "\n",
    "# variant calling options\n",
    "min_cov = 50 # minimum coverage \n",
    "min_var_freq = 0\n",
    "min_var_prop = 0.02 # minimum variant proportion \n",
    "err_tol = 0.01 # threshold to which variant called could result from base calling error \n",
    "min_breadth = 0.7 # min breadth of gene segment to be mapped for further analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Initialising CDS coordinates...\n",
      "\n",
      "Check translated protein sequences...\n",
      "PB2 MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMRYPITADKRIMDMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDTNPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLAITKEKKEELQDCKIAPLMVAYMLERELVRKTRFLPVAGGTGSVYIEVLHLTQGTCWEQMYTPGGEVRNDDVDQSLIIAARNIVRRAAVSADPLASLLEMCHSTQIGGVRMVDILRQNPTEEQAVDICKAAIGLRISSSFSFGGFTFKRTSGSSVKKEEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGRDEQSIAEAIIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIESIDNVMGMIGILPDMTPSTEMSLRGIRVSKMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTEKLTITYSSSMMWEINGPESVLVNTYQWIIRNWEIVKIQWSQDPTMLYNKMEFEPFQSLVPKATRSRYSGFVRTLFQQMRDVLGTFDTVQIIKLLPFAAAPPEQSRMQFSSLTVNVRGSGLRILVRGNSPVFNYNKATKRLTVLGKDAGALTEDPDEGTSGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*\n",
      "PB1 MDVNPTLLFLKIPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLETMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTIEVFRSNGLTANESGRLIDFLKDVMESMNKEEIEITTHFQRKRRVRDNMTKKMVTQRTIGKKKQRLNKRGYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTEISFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNILSMAPIMFSNKMARLGKGYMFESKRMKIRTQIPAEMLASIDLKYFNESTKKKIEKIRPLLIDGTASLSPGMMMGMFNMLSTVLGVSILNLGQKKYTKTIYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINKTGTFEFTSFFYRYGFVANFSMELPSFGVSGVNESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWDQTQSKVGLLVSDGGPNLYNIRNLHIPEVCLKWELMDDDYRGRLCNPLNPFVSHKEIDSVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCNLFEKFFPSSSYRRPVGISSMVEAMVSRARIDARVDFESGRIKKEEFSEIMKICSTIEELRRQK*\n",
      "PA MEDFVRQCFNPMIVELAEKAMKEYGEDPKIETNKFAAICTHLEVCFMYSDFHFIDERGESIIVESGDPNALLKHRFEIIEGRDRIMAWTVVNSICNTTGVEKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSEKTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMASRSLWDSFRQSERGEETIEEKFEITGTMRKLADQSLPPNFPSLENFRAYVDGFEPNGCIEGKLSQMSKEVNAKIEPFLRTTPRPLRLPDGPLCHQRSKFLLMDALKLSIEDPSHEGEGIPLYDAIKCMKTFFGWKEPNIVKPHEKGINPNYLMAWKQVLAELQDIENEEKIPRTKNMKRTSQLKWALGENMAPEKVDFDDCKDVGDLKQYDSDEPEPRSLASWVQNEFNKACELTDSSWIELDEIGEDVAPIEHIASMRRNYFTAEVSHCRATEYIMKGVYINTALLNASCAAMDDFQLIPMISKCRTKEGRRKTNLYGFIIKGRSHLRNDTDVVNFVSMEFSLTDPRLEPHKWEKYCVLEIGDMLLRTAIGQVSRPMFLYVRTNGTSKIKMKWGMEMRRCLLQSLQQIESMIEAESSVKEKDMTKEFFENKSETWPIGESPRGVEEGSIGKVCRTLLAKSVFNSLYASPQLEGFSAESRKLLLIVQALRDNLEPGTFDLGGLYEAIEECLINDPWVLLNASWFNSFLTHALK*\n",
      "PA-X MEDFVRQCFNPMIVELAEKAMKEYGEDPKIETNKFAAICTHLEVCFMYSDFHFIDERGESIIVESGDPNALLKHRFEIIEGRDRIMAWTVVNSICNTTGVEKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSEKTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMASRSLWDSFVSPKEAKRQLKKNLRLQELCASLPTKVSHRTSPALKTLEPM*\n",
      "HA MKAILVVLLYTFATANADTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTASSWSYIVETPSSDNGTCYPGDFIDYEELREQLSSVSSFERFEIFPKTSSWPNHDSNKGVTAACPHAGAKSFYKNLIWLVKKGNSYPKLSKSYINDKGKEVLVLWGIHHPSTSADQQSLYQNADTYVFVGSSRYSKKFKPEIAIRPKVRDQEGRMNYYWTLVEPGDKITFEATGNLVVPRYAFAMERNAGSGIIISDTPVHDCNTTCQTPKGAINTSLPFQNIHPITIGKCPKYVKSTKLRLATGLRNIPSIQSRGLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDEITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENLNKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRSQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYPKYSEEAKLNREEIDGVKLESTRIYQILAIYSTVASSLVLVVSLGAISFWMCSNGSLQCRICI*\n",
      "NP MASQGTKRSYEQMETGGERQDATEIRASVGRMIGGIGRFYIQMCTELKLSDYDGRLIQNSITIERMVLSAFDERRNKYLEEHPSAGKDPKKTGGPIYRRVDGKWMRELILYDKEEIRRVWRQANNGEDATAGLTHIMIWHSNLNDATYQRTRALVRTGMDPRMCSLMQGSTLPRRSGAAGAAVKGVGTIAMELIRMIKRGINDRNFWRGENGRRTRVAYERMCNILKGKFQTAAQRAMMDQVRESRNPGNAEIEDLIFLARSALILRGSVAHKSCLPACVYGLAVASGHDFEREGYSLVGIDPFKLLQNSQVVSLMRPNENPAHKSQLVWMACHSAAFEDLRVSSFIRGKKVIPRGKLSTRGVQIASNENVETMDSNTLELRSRYWAIRTRSGGNTNQQKASAGQISVQPTFSVQRNLPFERATVMAAFSGNNEGRTSDMRTEVIRMMESAKPEDLSFQGRGVFELSDEKATNPIVPSFDMSNEGSYFFGDNAEEYDS*\n",
      "NA MNPNQKIITIGSVCMTIGMANLILQIGNIISIWISHSIQLGNQNQIETCNQSVITYENNTWVNQTYVNISNTNFAAGQSVVSVKLAGNSSLCPVSGWAIYSKDNSVRIGSKGDVFVIREPFISCSPLECRTFFLTQGALLNDKHSNGTIKDRSPYRTLMSCPIGEVPSPYNSRFESVAWSASACHDGINWLTIGISGPDNGAVAVLKYNGIITDTIKSWRNNILRTQESECACVNGSCFTVMTDGPSNGQASYKIFRIEKGKIVKSVEMNAPNYHYEECSCYPDSSEITCVCRDNWHGSNRPWVSFNQNLEYQIGYICSGIFGDNPRPNDKTGSCGPVSSNGANGVKGFSFKYGNGVWIGRTKSISSRNGFEMIWDPNGWTGTDNNFSIKQDIVGINEWSGYSGSFVQHPELTGLDCIRPCFWVELIRGRPKENTIWTSGSSISFCGVNSDTVGWSWPDGAELPFTIDK*\n",
      "M1 MSLLTEVETYVLSIIPSGPLKAEIAQRLESVFAGKNTDLEALMEWLKTRPILSPLTKGILGFVFTLTVPSERGLQRRRFVQNALNGNGDPNNMDRAVKLYKKLKREITFHGAKEVSLSYSTGALASCMGLIYNRMGTVTTEAAFGLVCATCEQIADSQHRSHRQMATTTNPLIRHENRMVLASTTAKAMEQMAGSSEQAAEAMEVANQTRQMVHAMRTIGTHPSSSAGLKDDLLENLQAYQKRMGVQMQRFK*\n",
      "M2 MSLLTEVETPTRSEWECRCSDSSDPLVIAANIIGILHLILWITDRLFFKCIYRRFKYGLKRGPSTEGVPESMREEYQQEQQSAVDVDDGHFVNIELE*\n",
      "NEP MDSNTMSSFQYLLFQDILMRMSKMQLGSSSEDLNGMVTRFESLKIYRDSLGETVMRMGDLHYLQSRNEKWREQLGQKFEEIRWLIEEMRHRLKATENSFEQITFMQALQLLLEVEQEIRAFSFQLI*\n",
      "NS1 MDSNTMSSFQVDCFLWHIRKRFADNGLGDAPFLDRLRRDQKSLKGRGNTLGLDIETATLVGKQIVEWILKEESSETLRMTIASVPTSRYLSDMTLEEMSRDWFMLMPRQKIIGPLCVRLDQAIMEKNIVLKANFSVIFNRLETLILLRAFTEEGAIVGEISPLPSLPGHTYEDVKNAVGVLIGGLEWNGNTVRVSENIQRFAWRNCDENGRPSLPPEQK*\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>frame</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>nuc</th>\n",
       "      <th>protein</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1-PB2</th>\n",
       "      <th>1</th>\n",
       "      <th>PB2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <th>PB2</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <th>PB2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <th>PB2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <th>PB2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8-NS</th>\n",
       "      <th>834</th>\n",
       "      <th>NEP</th>\n",
       "      <td>126</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <th>NEP</th>\n",
       "      <td>126</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <th>NEP</th>\n",
       "      <td>127</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <th>NEP</th>\n",
       "      <td>127</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838</th>\n",
       "      <th>NEP</th>\n",
       "      <td>127</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14106 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    aa  frame\n",
       "gene  nuc protein            \n",
       "1-PB2 1   PB2        1      1\n",
       "      2   PB2        1      2\n",
       "      3   PB2        1      3\n",
       "      4   PB2        2      1\n",
       "      5   PB2        2      2\n",
       "...                ...    ...\n",
       "8-NS  834 NEP      126      2\n",
       "      835 NEP      126      3\n",
       "      836 NEP      127      1\n",
       "      837 NEP      127      2\n",
       "      838 NEP      127      3\n",
       "\n",
       "[14106 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# presets \n",
    "reffasta = ref_fasta_fname\n",
    "\n",
    "# initialise\n",
    "gene_to_proteinorf, influenza_gene_len, sorted_refnames, nucpos_shift = ngs.initialisation(cds_coords, reffasta, laeb_lib, nucpos_shift=nucpos_shift)\n",
    "gene_to_proteinorf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata and variant calling files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>enrolD</th>\n",
       "      <th>household</th>\n",
       "      <th>date</th>\n",
       "      <th>project</th>\n",
       "      <th>short_name</th>\n",
       "      <th>CT</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>...</th>\n",
       "      <th>ost_ed</th>\n",
       "      <th>ost_days</th>\n",
       "      <th>1-PB2</th>\n",
       "      <th>2-PB1</th>\n",
       "      <th>3-PA</th>\n",
       "      <th>4-HA</th>\n",
       "      <th>5-NP</th>\n",
       "      <th>6-NA</th>\n",
       "      <th>7-M</th>\n",
       "      <th>8-NS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3-20</th>\n",
       "      <td>11-001</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13/11/09</td>\n",
       "      <td>Hanoi</td>\n",
       "      <td>11-001-d0</td>\n",
       "      <td>17.00</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1;11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.978012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.970606</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-22</th>\n",
       "      <td>11-002</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17/11/09</td>\n",
       "      <td>Hanoi</td>\n",
       "      <td>11-002-d0</td>\n",
       "      <td>17.28</td>\n",
       "      <td>18.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3;13</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985303</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-24</th>\n",
       "      <td>11-006</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12/11/09</td>\n",
       "      <td>Hanoi</td>\n",
       "      <td>11-006-d2</td>\n",
       "      <td>27.42</td>\n",
       "      <td>38.5</td>\n",
       "      <td>Male</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1;9</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-1</th>\n",
       "      <td>11-1015</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22/10/09</td>\n",
       "      <td>Hanoi</td>\n",
       "      <td>11-1015-d0</td>\n",
       "      <td>25.00</td>\n",
       "      <td>39.4</td>\n",
       "      <td>Female</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3;7</td>\n",
       "      <td>0.945175</td>\n",
       "      <td>0.956025</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-3</th>\n",
       "      <td>11-1015</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27/10/09</td>\n",
       "      <td>Hanoi</td>\n",
       "      <td>11-1015-d4</td>\n",
       "      <td>33.09</td>\n",
       "      <td>39.4</td>\n",
       "      <td>Female</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3;7</td>\n",
       "      <td>0.320175</td>\n",
       "      <td>0.098505</td>\n",
       "      <td>0.198047</td>\n",
       "      <td>0.764844</td>\n",
       "      <td>0.832999</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       subject_id  enrolD household      date project  short_name     CT  \\\n",
       "sampid                                                                     \n",
       "3-20       11-001       0       NaN  13/11/09   Hanoi   11-001-d0  17.00   \n",
       "3-22       11-002       0       NaN  17/11/09   Hanoi   11-002-d0  17.28   \n",
       "3-24       11-006       2       NaN  12/11/09   Hanoi   11-006-d2  27.42   \n",
       "3-1       11-1015       0       NaN  22/10/09   Hanoi  11-1015-d0  25.00   \n",
       "3-3       11-1015       4       NaN  27/10/09   Hanoi  11-1015-d4  33.09   \n",
       "\n",
       "         age     sex  timepoint  ... ost_ed ost_days     1-PB2     2-PB1  \\\n",
       "sampid                           ...                                       \n",
       "3-20    10.0  Female        2.0  ...   11.0     1;11  1.000000  0.978012   \n",
       "3-22    18.0  Female        4.0  ...   13.0     3;13  1.000000  1.000000   \n",
       "3-24    38.5    Male        4.0  ...    9.0      1;9  1.000000  1.000000   \n",
       "3-1     39.4  Female        2.0  ...    7.0      3;7  0.945175  0.956025   \n",
       "3-3     39.4  Female        7.0  ...    7.0      3;7  0.320175  0.098505   \n",
       "\n",
       "            3-PA      4-HA      5-NP  6-NA  7-M 8-NS  \n",
       "sampid                                                \n",
       "3-20    1.000000  0.970606  1.000000   1.0  1.0  1.0  \n",
       "3-22    1.000000  0.985303  1.000000   1.0  1.0  1.0  \n",
       "3-24    1.000000  1.000000  1.000000   1.0  1.0  1.0  \n",
       "3-1     1.000000  1.000000  1.000000   1.0  1.0  1.0  \n",
       "3-3     0.198047  0.764844  0.832999   1.0  1.0  1.0  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>nucpos</th>\n",
       "      <th>nuc_var</th>\n",
       "      <th>nuc_prop</th>\n",
       "      <th>nuc_freq</th>\n",
       "      <th>nuc_consensus</th>\n",
       "      <th>nuc_consensus_freq</th>\n",
       "      <th>nuc_coverage</th>\n",
       "      <th>protein</th>\n",
       "      <th>aapos</th>\n",
       "      <th>HA_num_type</th>\n",
       "      <th>aa_var</th>\n",
       "      <th>aa_prop</th>\n",
       "      <th>aa_freq</th>\n",
       "      <th>expected_aa_consensus</th>\n",
       "      <th>aa_consensus</th>\n",
       "      <th>aa_consensus_freq</th>\n",
       "      <th>aa_coverage</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>codon</th>\n",
       "      <th>nonsyn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">2-17</th>\n",
       "      <th>3-PA</th>\n",
       "      <td>204</td>\n",
       "      <td>A</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>77</td>\n",
       "      <td>91</td>\n",
       "      <td>PA</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>0.162791</td>\n",
       "      <td>14.0</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>72.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>3</td>\n",
       "      <td>CCA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-PA</th>\n",
       "      <td>204</td>\n",
       "      <td>A</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>77</td>\n",
       "      <td>91</td>\n",
       "      <td>PA-X</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>0.162791</td>\n",
       "      <td>14.0</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>72.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>3</td>\n",
       "      <td>CCA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-PA</th>\n",
       "      <td>1321</td>\n",
       "      <td>C</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "      <td>72</td>\n",
       "      <td>84</td>\n",
       "      <td>PA</td>\n",
       "      <td>441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>0.148148</td>\n",
       "      <td>12.0</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>69.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CTG</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-PA</th>\n",
       "      <td>1848</td>\n",
       "      <td>A</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>80</td>\n",
       "      <td>92</td>\n",
       "      <td>PA</td>\n",
       "      <td>616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>0.134831</td>\n",
       "      <td>12.0</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>77.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>3</td>\n",
       "      <td>TCA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4-HA</th>\n",
       "      <td>249</td>\n",
       "      <td>A</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>64</td>\n",
       "      <td>67</td>\n",
       "      <td>HA</td>\n",
       "      <td>75</td>\n",
       "      <td>H3</td>\n",
       "      <td>E</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>3.0</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>64.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GAA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             nucpos nuc_var  nuc_prop  nuc_freq nuc_consensus  \\\n",
       "sampid gene                                                     \n",
       "2-17   3-PA     204       A  0.153846        14             G   \n",
       "       3-PA     204       A  0.153846        14             G   \n",
       "       3-PA    1321       C  0.142857        12             A   \n",
       "       3-PA    1848       A  0.130435        12             G   \n",
       "       4-HA     249       A  0.044776         3             G   \n",
       "\n",
       "             nuc_consensus_freq  nuc_coverage protein  aapos HA_num_type  \\\n",
       "sampid gene                                                                \n",
       "2-17   3-PA                  77            91      PA     68         NaN   \n",
       "       3-PA                  77            91    PA-X     68         NaN   \n",
       "       3-PA                  72            84      PA    441         NaN   \n",
       "       3-PA                  80            92      PA    616         NaN   \n",
       "       4-HA                  64            67      HA     75          H3   \n",
       "\n",
       "            aa_var   aa_prop  aa_freq expected_aa_consensus aa_consensus  \\\n",
       "sampid gene                                                                \n",
       "2-17   3-PA      P  0.162791     14.0                     P            P   \n",
       "       3-PA      P  0.162791     14.0                     P            P   \n",
       "       3-PA      L  0.148148     12.0                     M            M   \n",
       "       3-PA      S  0.134831     12.0                     S            S   \n",
       "       4-HA      E  0.044776      3.0                     E            E   \n",
       "\n",
       "             aa_consensus_freq  aa_coverage  codon_pos codon  nonsyn  \n",
       "sampid gene                                                           \n",
       "2-17   3-PA               72.0         86.0          3   CCA     0.0  \n",
       "       3-PA               72.0         86.0          3   CCA     0.0  \n",
       "       3-PA               69.0         81.0          1   CTG     1.0  \n",
       "       3-PA               77.0         89.0          3   TCA     0.0  \n",
       "       4-HA               64.0         67.0          3   GAA     0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# metadata \n",
    "## metadata must have 'sampid' header which is used as sample identifier \n",
    "meta_df = pd.read_csv('%s'%meta_fname)\n",
    "sorted_sampid = sorted(set(meta_df['sampid']))\n",
    "\n",
    "meta_df = meta_df.set_index(\"sampid\")\n",
    "display (meta_df.head())\n",
    "\n",
    "# vcf \n",
    "variant_call_df = pd.read_csv(\"./results/variant_call_MinCoV%i_MinProp%.2f_MinFreq%i_ErrTol%.2f.csv\"%(min_cov, min_var_prop, min_var_freq, err_tol))\n",
    "variant_call_df = variant_call_df.set_index([\"sampid\", 'gene'])\n",
    "display (variant_call_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get consensus and putative minority haplotype sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampid</th>\n",
       "      <th>gene</th>\n",
       "      <th>seqtype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">2-18</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">8-NS</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGGACTCCAACACCATGTCAAGCTTTCAGGTAGACTGTTTCCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>ATGGACTCCAACACCATGTCAAGCTTTCAGGTAGACTGTTTCCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">5-NP</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>ATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7-M</th>\n",
       "      <th>C</th>\n",
       "      <td>ATGAGTCTTCTAACCGAGGTCGAAACGTACGTTCTTTCTATCATCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   seq\n",
       "sampid gene seqtype                                                   \n",
       "2-18   8-NS C        ATGGACTCCAACACCATGTCAAGCTTTCAGGTAGACTGTTTCCTTT...\n",
       "            M1       ATGGACTCCAACACCATGTCAAGCTTTCAGGTAGACTGTTTCCTTT...\n",
       "       5-NP C        ATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTG...\n",
       "            M1       ATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTG...\n",
       "       7-M  C        ATGAGTCTTCTAACCGAGGTCGAAACGTACGTTCTTTCTATCATCC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if not os.path.isdir(\"./reference\"): \n",
    "    os.mkdir(\"./reference\")\n",
    "\n",
    "def get_consensus_minority_seq(sampid):\n",
    "    gene_to_consensus_seq_row = []\n",
    "    \n",
    "    # check that all gene segment satisfy min_breadth \n",
    "    if (meta_df.loc[sampid, sorted_refnames]>min_breadth).all() == False: \n",
    "        return gene_to_consensus_seq_row\n",
    "\n",
    "    # read map_nuc_results \n",
    "    map_nuc_results = pd.read_csv('./results/map_nuc_results_%s.csv'%(sampid))\n",
    "    map_nuc_results = map_nuc_results.set_index([\"Gene\", \"Position\"])\n",
    "\n",
    "    # sample_vcf \n",
    "    try:\n",
    "        sample_vcf = variant_call_df.loc[sampid]\n",
    "    except: \n",
    "        sample_vcf = None\n",
    "\n",
    "    for gene in set(map_nuc_results.index.get_level_values(0)): \n",
    "\n",
    "        # add consensus sequence \n",
    "        sequence = {position:map_nuc_results.loc[(gene, position), \"Consensus\"] if (gene, position) in list(map_nuc_results.index) else \"-\" for position in range(1, influenza_gene_len[gene]+1)}\n",
    "        conseq = \"\".join([sequence[pos] for pos in sorted(sequence.keys())])\n",
    "        gene_to_consensus_seq_row.append({\"sampid\":sampid, \"gene\":gene, \"seq\":conseq, \"seqtype\":\"C\"})\n",
    "\n",
    "        # add putative minority haplotype sequence \n",
    "        try: \n",
    "            gene_sample_vcf = sample_vcf.loc[gene].copy()\n",
    "        except: \n",
    "            continue \n",
    "\n",
    "        if isinstance(gene_sample_vcf, pd.Series):\n",
    "            gene_sample_vcf = gene_sample_vcf.to_frame().T\n",
    "\n",
    "        gene_sample_vcf = gene_sample_vcf.drop_duplicates([\"nucpos\", \"nuc_var\"])\n",
    "        gene_sample_vcf = gene_sample_vcf.set_index(['nucpos'])\n",
    "        \n",
    "        pos_to_var = {}\n",
    "        for pos in list(gene_sample_vcf.index):\n",
    "            nuc_var = gene_sample_vcf.loc[pos, \"nuc_var\"]\n",
    "            if isinstance(nuc_var, str): # single variant \n",
    "                pos_to_var[pos] = [gene_sample_vcf.loc[pos, \"nuc_var\"]]\n",
    "            else: # multi variant \n",
    "                pos_to_var[pos] = list(gene_sample_vcf.loc[pos, 'nuc_var'])\n",
    "        \n",
    "        minseq = [pos_to_var[pos] if pos in pos_to_var else [sequence[pos]] for pos in range(1, influenza_gene_len[gene]+1)]\n",
    "        for midx, mseq in enumerate(list(itertools.product(*minseq))): \n",
    "            gene_to_consensus_seq_row.append({\"sampid\":sampid, \"gene\":gene, \"seq\":\"\".join(mseq), \"seqtype\":\"M%i\"%(midx+1)})\n",
    "        \n",
    "    return gene_to_consensus_seq_row\n",
    "\n",
    "if os.path.isfile(\"./reference/consensus_sequences.csv\"):\n",
    "    gene_to_consensus_seq = pd.read_csv(\"./reference/consensus_sequences.csv\")\n",
    "else:\n",
    "    \"\"\"for sampid in sorted_sampid:\n",
    "        get_consensus_minority_seq(sampid)\"\"\"\n",
    "    pool = mp.Pool(processes=threadnum)\n",
    "    results = [pool.apply_async(get_consensus_minority_seq, args=(sampid,)) for sampid in sorted_sampid]\n",
    "    output = [p.get() for p in results]\n",
    "\n",
    "    gene_to_consensus_seq = [k for v in list(filter(None, output)) for k in v]\n",
    "    gene_to_consensus_seq = pd.DataFrame.from_dict(gene_to_consensus_seq)\n",
    "\n",
    "    # save to file \n",
    "    gene_to_consensus_seq.to_csv(\"./reference/consensus_sequences.csv\", index=False)\n",
    "\n",
    "gene_to_consensus_seq = gene_to_consensus_seq.set_index([\"sampid\", \"gene\", \"seqtype\"])\n",
    "display (gene_to_consensus_seq.head()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenate consensus sequences of all gene segments "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome = fc.parsefasta(reffasta)\n",
    "\n",
    "# write fasta \n",
    "if os.path.isfile(\"./reference/concatenated_wgs.fasta\") == False:\n",
    "    with open(\"./reference/concatenated_wgs.fasta\", \"w\") as output: \n",
    "        output.write(\">%s\\n%s\\n\"%(re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1]),\n",
    "                                  \"\".join([reference_genome[gene] for gene in sorted_refnames]).upper()))\n",
    "\n",
    "        for sampid in set(gene_to_consensus_seq.index.get_level_values(0)): \n",
    "            #  write concatenated consensus sequence \n",
    "            sample_gene_to_consensus_seq = gene_to_consensus_seq.loc[sampid]\n",
    "\n",
    "            header = \"%s|%s\"%(sampid, \"C\")\n",
    "            sequence =  \"\".join([sample_gene_to_consensus_seq.loc[(gene, \"C\"), \"seq\"] for gene in sorted_refnames])\n",
    "            output.write(\">%s\\n%s\\n\"%(header, sequence))\n",
    "\n",
    "            # write putative minority haplotype sequence \n",
    "            for seqtype in set(sample_gene_to_consensus_seq.index.get_level_values(1)): \n",
    "                if seqtype == \"C\": \n",
    "                    continue \n",
    "\n",
    "                header = \"%s|%s\"%(sampid, seqtype)\n",
    "                sequence =  []\n",
    "                for gene in sorted_refnames: \n",
    "                    if (gene, seqtype) in sample_gene_to_consensus_seq.index: \n",
    "                        sequence.append(sample_gene_to_consensus_seq.loc[(gene, seqtype), \"seq\"])\n",
    "                    else: \n",
    "                        sequence.append(sample_gene_to_consensus_seq.loc[(gene, \"C\"), \"seq\"])\n",
    "                output.write(\">%s\\n%s\\n\"%(header, \"\".join(sequence))) \n",
    "\n",
    "if os.path.isfile(\"./reference/mafft_concatenated_wgs.fasta\") == False:\n",
    "    cmd = [\"mafft\", \"./reference/concatenated_wgs.fasta\", \">\", \"./reference/mafft_concatenated_wgs.fasta\"]\n",
    "    subprocess.call(\" \".join(cmd), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruct phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build tree \n",
    "if os.path.isfile(\"./reference/concatenated_wgs.fasta.treefile\") == False:\n",
    "    cmd = ['iqtree', '-nt', 'AUTO', '-ntmax', str(threadnum), \n",
    "           '-m', 'GTR+I+G4', \n",
    "           '-s', './reference/mafft_concatenated_wgs.fasta']\n",
    "    subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate newick tree for printing and annotation dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate ggtree dataframe and newick tree file\n",
    "# get treetime annotations \n",
    "tree_annotations = []\n",
    "\n",
    "ete3_tree = ete3.Tree(\"./reference/mafft_concatenated_wgs.fasta.treefile\")\n",
    "ete3_tree.set_outgroup(re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1]))\n",
    "ete3_tree.ladderize()\n",
    "\n",
    "for leaf in ete3_tree.get_leaves():\n",
    "    # skip vaccine root \n",
    "    if (leaf.name == re.sub(\"\\.fasta$\", \"\", ref_fasta_fname.split(\"/\")[-1])): \n",
    "        print_name = leaf.name\n",
    "        subject_id = \"\"\n",
    "        seq_type = \"Consensus\" \n",
    "    else: \n",
    "        sampid, stype = leaf.name.split(\"|\")\n",
    "        subject_id = meta_df.loc[sampid, \"subject_id\"]\n",
    "        timepoint = meta_df.loc[sampid, \"timepoint\"]\n",
    "        \"\"\"print (subject_id, timepoint, stype)\n",
    "        continue\"\"\"\n",
    "\n",
    "        timepoint = meta_df.loc[sampid, \"timepoint\"]\n",
    "        if pd.isna(timepoint): \n",
    "            timepoint = \"D-NaN\"\n",
    "        else: \n",
    "            timepoint = \"D%i\"%(int(timepoint))\n",
    "\n",
    "        if stype == \"C\": \n",
    "            seq_type = \"Consensus\"\n",
    "        else: \n",
    "            if subject_id == \"H296/S03\" or subject_id == \"H089/S04\":\n",
    "                minority_idx = int(re.search(\"M(\\d+)$\", leaf.name).group(1))\n",
    "                seq_type = \"Minor-%i\"%(minority_idx)\n",
    "            else:\n",
    "                seq_type = \"Minor\"\n",
    "\n",
    "        print_name = \"_\".join([subject_id, timepoint, seq_type])\n",
    "        \n",
    "        \n",
    "    tree_annotations.append({\"index\":leaf.name, \"print_name\":print_name, \n",
    "                             \"show_name\":1, \n",
    "                             \"show_tip\":1 if seq_type == \"Consensus\" else 0, \n",
    "                             \"subject_id\":subject_id})\n",
    "\n",
    "# save for ggtree use \n",
    "tree_annotations = pd.DataFrame.from_dict(tree_annotations).set_index(\"index\")\n",
    "tree_annotations.to_csv(\"./reference/ggtree_concatenated_wgs.meta.csv\")\n",
    "ete3_tree.write(outfile=\"./reference/ggtree_concatenated_wgs.nwk\", format=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
